tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_execution_payload_header.py

```diff
@@ -0,0 +1,463 @@
+    """
+    Test invalid signature fails
+    """
+    proposer_index = spec.get_beacon_proposer_index(state)
+
+    block, signed_header = prepare_block_with_execution_payload_header(
+        spec, state, builder_index=proposer_index, valid_signature=False
+    )
+
+    yield from run_execution_payload_header_processing(spec, state, block, valid=False)
+
+
+#
+# Builder validation tests
+#
+
+
+@with_eip7732_and_later
+@spec_state_test
+def test_process_execution_payload_header_inactive_builder(spec, state):
+    """
+    Test inactive builder fails
+    """
+    # Make builder inactive by setting exit epoch
+    builder_index = spec.get_beacon_proposer_index(state)
+    state.validators[builder_index].exit_epoch = spec.get_current_epoch(state)
+
+    make_validator_builder(spec, state, builder_index)
+
+    block, signed_header = prepare_block_with_execution_payload_header(
+        spec, state, builder_index=builder_index, value=spec.Gwei(1000000)
+    )
+
+    yield from run_execution_payload_header_processing(spec, state, block, valid=False)
+
+
+@with_eip7732_and_later
+@spec_state_test
+def test_process_execution_payload_header_slashed_builder(spec, state):
+    """
+    Test slashed builder fails
+    """
+    builder_index = spec.get_beacon_proposer_index(state)
+
+    # Slash the builder
+    state.validators[builder_index].slashed = True
+    make_validator_builder(spec, state, builder_index)
+
+    block, signed_header = prepare_block_with_execution_payload_header(
+        spec, state, builder_index=builder_index, value=spec.Gwei(1000000)
+    )
+
+    yield from run_execution_payload_header_processing(spec, state, block, valid=False)
+
+
+@with_eip7732_and_later
+@spec_state_test
+def test_process_execution_payload_header_non_builder_non_zero_value(spec, state):
+    """
+    Test non-builder attempting non-zero value fails
+    """
+    proposer_index = spec.get_beacon_proposer_index(state)
+
+    # Don't make proposer a builder, but try non-zero value
+    block, signed_header = prepare_block_with_execution_payload_header(
+        spec,
+        state,
+        builder_index=proposer_index,
+        value=spec.Gwei(1000000),  # Non-zero value should fail for non-builder
+    )
+
+    yield from run_execution_payload_header_processing(spec, state, block, valid=False)
+
+
+@with_eip7732_and_later
+@spec_state_test
+def test_process_execution_payload_header_non_builder_wrong_proposer(spec, state):
+    """
+    Test non-builder with wrong proposer index fails
+    """
+    proposer_index = spec.get_beacon_proposer_index(state)
+    other_index = (proposer_index + 1) % len(state.validators)
+
+    # Non-builder but not the proposer
+    block, signed_header = prepare_block_with_execution_payload_header(
+        spec, state, builder_index=other_index, value=spec.Gwei(0)
+    )
+
+    yield from run_execution_payload_header_processing(spec, state, block, valid=False)
+
+
+#
+# Balance validation tests
```

Comment:
Another balance test I'd like to see:

* There is a builder with an excess balance, eg 2048.25 ETH.
* It is able to submit a bid for 2016.25 ETH.

Sort of an edge case because the bid limit depends on the builder's balance, not its effective balance.

Reply:
added in #[4f2db69](https://github.com/ethereum/consensus-specs/pull/4457/commits/4f2db6950780bf31f4b53c922d0c7f8c6ce9053c)